ocaml/xapi/vm_evacuation.ml

```
@@ -0,0 +1,101 @@
+       evacuation failed *)
+    Tasks.with_tasks_destroy ~rpc ~session_id ~timeout ~tasks |> ignore
+  in
+
+  let clean_shutdown vms =
+    TaskHelper.exn_if_cancelling ~__context; (* First check if _we_ have been cancelled *)
+    let tasks =
+      vms
+      |> List.filter (fun vm ->
+             List.mem `clean_shutdown (Client.VM.get_allowed_operations ~rpc ~session_id ~self:vm))
+      |> List.map (fun vm ->
+             let name_label = Client.VM.get_name_label ~rpc ~session_id ~self:vm in
+             debug "Requesting clean shutdown of VM: %s" name_label;
+             Client.Async.VM.clean_shutdown ~rpc ~session_id ~vm) in
+    Tasks.with_tasks_destroy ~rpc ~session_id ~timeout:60. ~tasks |> ignore
+  in
+
+  let hard_shutdown vms =
+    TaskHelper.exn_if_cancelling ~__context; (* First check if _we_ have been cancelled *)
+    let tasks =
+      vms
+      |> List.map (fun vm ->
+             let name_label = Client.VM.get_name_label ~rpc ~session_id ~self:vm in
+             debug "Requesting hard shutdown of VM: %s" name_label;
+             Client.Async.VM.hard_shutdown ~rpc ~session_id ~vm) in
+    (* no timeout: we need the VMs to be off *)
+    Tasks.wait_for_all ~rpc ~session_id ~tasks;
+    vms
+    |> List.filter is_running
+    |> List.iter (fun vm ->
+           let name_label = Client.VM.get_name_label ~rpc ~session_id ~self:vm in
+           info "Failure performing hard shutdown of VM: %s" name_label)
+  in
+
+  let shutdown vms =
+    log_and_ignore_exn (fun () -> clean_shutdown vms);
+    (* XXX: shutdown.py only acted on running domains, but what about
+     * suspended/paused? *)
+    vms
+    |> List.filter is_running
+    |> hard_shutdown
+  in
+
+  Client.Host.get_vms_which_prevent_evacuation ~rpc ~session_id ~self:host
+  |> Xapi_stdext_std.Listext.List.filter_map (fun (vm, _) ->
+         if self_managed_poweroff vm then None
+         else Some vm)
+  |> shutdown;
```

Maybe we should ignore exceptions here?
Or something like:
```
ignore_exn( shutdown vms which prevent evacuation; evacuate);
ignore_exn( shutdown rest of the vms )
```

---

Yes, we could wrap lines 82-86 into a `log_and_ignore_exn`.

---

Done